Add automatic rotation of images that contain orientation data

runfalk opened the pull request:
I have added automatic rotation when decoding a `RawImage`. Right now it is implemented as a separate image operation rather than being part of the demosaic step. I have experimented with in-situ rotation vs creating a new `OpBuffer` and found creating a new one to be more performant due to that it is easy to use `rayon`'s parallell iterators.

Criticism is welcome.

pedrocr commented on src/decoders/arw.rs:
```diff
@@ -45,6 +45,14 @@ impl<'a> Decoder for ArwDecoder<'a> {
     };
     let src = &self.buffer[offset..];
 
+    // Lookup the orientation of the image to store for rotating after demosaic
```
This code seems to be duplicated across formats. Almost all formats should use the same TIFF tag with the same interpretation so it should be possible to refactor this into a single central bit of code and then allow only the weird formats override like we already do for a few other bits of metadata.

runfalk replied:
I will refactor these into a separate function. I found out now that Canon has a tag called "Camera Orientation" that supports: Horizontal (normal), 1 = Rotate 90 CW and 2 = Rotate 270 CW. I assume they are always the same as Orientation and I guess they are part of the maker note. Should vendor specific ones take precedence or be used only if the standard orientation tag isn't available? I don't know if other vendors have similar tags.

pedrocr replied:
There's alot of duplication in tags. Most of the time the standard TIFF orientation tag should work. Probably all the time in case of Orientation as I think any format worth caring about implements it. dcraw as usual handles a bunch of the oddball ones but for this kind of thing they can just safely be ignored. Orientation is easy to fix manually if you actually do have one of the weirder formats.

As for the refactor I would try to make it so most formats don't even need any orientation code at all. check_supported() could probably stuff that info into the Camera struct and ok_image pick it up automatically. That way only formats that handle things more manually (like DNG or MRW) need to specifically call your orientation finding function.

pedrocr commented on src/decoders/mod.rs:
```diff
@@ -186,12 +186,43 @@ impl Camera {
-  Ok(RawImage::new(camera, width, height, wb_coeffs, image))
+/// Possible values for the IFD tag Orientation (0x0112)
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Orientation {
```
Does the TIFF tag really allow all these variations? I've only seen the 4 rotations in practice.

runfalk replied:
I just tried some of the uncommon ones in Adobe CameraRaw and macOS preview. Both of them respected `IFD0:Orientation` for both CR2 and TIFF. I tried both vertical and horisontal flipping. As for a camera actually producing images with these values I kind of doubt it.

I think it makes little sense to not allow it, as the functionality is already there and it doesn't make the handling any more complicated.

pedrocr replied:
Yeah, I'm sure TIFF allows all the options. I doubt any camera uses them but it doesn't matter really.

pedrocr commented on src/imageops/transform.rs:
```diff
@@ -0,0 +1,65 @@
+pub fn rotate(img: &RawImage, buf: &OpBuffer) -> OpBuffer {
+  match img.orientation {
+    Orientation::HorizontalFlip => flip_horizontal(buf),
+    Orientation::Rotate180 => flip_horizontal(&flip_vertical(buf)),
```
Won't things like this do multiple passes of processing? It's probably possible to do this in a single pass and gain some performance no? Does it make the code too ugly?

runfalk replied:
Yes. I actually tried an in-place rotation algorithm directly translated from some text file I found in the matrix library numpy for Python (https://github.com/numpy/numpy/blob/master/doc/f2py/multiarray/transpose.txt#L937). The solution was kind of slow and as far as I can tell not possible to run in parallel. Even a two-pass flipping operation ended up being faster as each operation is tiny in comparison and it could easily run in parallel using rayon. Though I don't remember the exact numbers I think they dropped from ~1300 ms to ~500 ms for a full resolution RAW image.

I think this is enough as a first implementation that could be improved later on if we can find a faster implementation.

pedrocr replied:
I mean something like this:

https://github.com/pedrocr/dcraw/blob/master/dcraw.c#L9824-L9836

I think any orientation can be described as a combination of swap_xy/flip_x/flip_y and a single loop can be written that does the right thing in whichever way the 3 bits are set. That should be quite fast and hopefully also easy to run in parallel.

runfalk replied:
I will look into it and benchmark to see if I can make it faster. The code seems simple enough.

pedrocr replied:
It's not a worry for now. If you can refactor the code duplication this is mostly fine to merge.

pedrocr replied:
I had a bit of time to think about this and it doesn't seem too hard to do it in one go. Here's some example code that should work (not even compile tested yet):

https://gist.github.com/pedrocr/4e65061da3dfa1af5b2c2fe2ac5e149e

This kind of code is tricky enough that it makes sense to have some unit tests for it. It should be simple enough to have some 3x4 pixel OpBuffer test examples for each of the orientation cases.

runfalk replied:
Would it be possible to leave this for another pull request? I can have a look at it once this one is merged.

pedrocr commented on src/imageops/mod.rs:
```diff
@@ -61,6 +63,11 @@ pub fn simple_decode (img: &RawImage, maxwidth: usize, maxheight: usize) -> OpBu
 
+  // Fix orientation if necessary and possible
+  if img.orientation != Orientation::Normal && img.orientation != Orientation::Unknown {
+    channel4 = do_timing("rotate", || { transform::rotate(img, &channel4) });
```
When there is a transpose this will break the maxwidth/maxheight contract. We'll need proper pipeline sizing code at some point but for now just detecting those cases and flipping maxwidth/maxheight is enough.

pedrocr commented on src/decoders/arw.rs:
```diff
@@ -75,6 +67,7 @@ impl<'a> Decoder for ArwDecoder<'a> {
       },
       _ => return Err(format!("ARW: Don't know how to decode type {}", compression).to_string()),
     };
+    let orientation = Orientation::from_tiff_ifd(&self.tiff);
```
I'd prefer if you got rid of these lines completely. Just stuff this call into check_supported() and push orientation into the Camera struct so that ok_image() can access it.

runfalk replied:
Isn't this a problem since the `Camera` instances are stored in a `HashMap` with no other image specific data? Doesn't two images from the same camera share the same `Camera` instance? Do we want another metadata struct instead?

pedrocr replied:
We just need to clone the camera instead of returning the one in the HashMap. The one in the hash is the one with the defaults, the clone is the one that we alter with the metadata from this camera.

pedrocr commented on src/decoders/mod.rs:
```diff
@@ -186,12 +186,66 @@ impl Camera {
+
+  /// Return true if orientation results in a transpose, i.e. width and height
+  /// are swapped
+  pub fn is_transpose(&self) -> bool {
```
This seems like a code smell. It probably makes more sense to make the enum generic as a (bool, bool, bool) of (swapxy, flipx, flipy) and then this comes for free and those same three bools can be passed to a swapflip style function like the one I described.

runfalk replied:
I'd argue that keeping the current enum is clearer since it is easy to interpret the meanings of the orientations. Do you suggest something like `Orientation::into::<ImageTransformation>()` where `ImageTransformation` (or a better name) would contain `transpose`, `flip_x` and `flip_y` as fields? That way you could pass either into a swapflip style function and it would work.

pedrocr replied:
My knowledge of rust enums is still quite incomplete. Can't you have the same named enums but have them be instances of the same generic 3 bool item? If not the conversion one seems fine. I'd call it something like OrientationFlips instead of ImageTransformation.

runfalk replied:
I've tried to dive a bit deeper into enums and as far as I can tell the `::into()` method seems the most "rustic"(?). I think I'm gonna go with that.

pedrocr replied:
Sounds good. I asked around on IRC and the consensus was the same

runfalk commented:
All right. I have changed things around and put orientation onto the `Camera` object as well. Since we are now cloning them, I reworked the explicit lifetime `'a` for them. The `ok_image`, `RawImage` and their likes now consumes a `Camera` instance instead of holding onto a reference.

I think this is a better solution since explicit lifetimes can be avoided, though I don't know what the original intention behind passing a reference around was.

pedrocr commented on src/decoders/mod.rs:
```diff
@@ -182,15 +184,94 @@ impl Camera {
+impl Into<OrientationFlips> for Orientation {
+  fn into(self) -> OrientationFlips {
+    OrientationFlips {
+      transpose: match self {
```
Wouldn't it be simpler to just do a single match on all the orientation flips and do OrientationFlips{bool,bool,bool} for each of them?

pedrocr commented on src/decoders/mrw.rs:
```diff
@@ -46,7 +46,7 @@ impl<'a> MrwDecoder<'a> {
           }
         }
         0x545457 => { // TTW
-          // Base value for offsets needs to be at the beginning of the 
```
Please avoid making whitespace changes in unrelated PRs.

runfalk replied:
Whoops, it seems like I forgot to turn off my editor's trim whitespace on save feature.

runfalk replied:
Turns out most of these whitespace changes were introduced in the initial commit. When the rest is done I will squash all the commits and fix the whitespace changes.

pedrocr commented on src/decoders/tiff.rs:
```diff
@@ -91,6 +92,7 @@ pub enum Tag {
   KdcIFD           = 0xFE00,
 }
 }
+
```
Another one

pedrocr commented on src/decoders/mrw.rs:
```diff
@@ -55,7 +55,7 @@ impl<'a> MrwDecoder<'a> {
     }
 
-    MrwDecoder { 
+    MrwDecoder {
```
Another

pedrocr commented on src/decoders/nef.rs:
```diff
@@ -147,6 +147,7 @@ impl<'a> Decoder for NefDecoder<'a> {
       img.blacklevels = [0,0,0,0];
       img.whitelevels = [65535,65535,65535,65535];
     }
+
```
and another :)

pedrocr commented on src/imageops/transform.rs:
```diff
@@ -0,0 +1,64 @@
+/// Rotate an OpBuffer based on the given RawImage's orientation
+pub fn rotate(img: &RawImage, buf: &OpBuffer) -> OpBuffer {
+  match img.orientation.into() {
+    OrientationFlips { transpose: false, flip_x: false, flip_y: false } => buf.clone(),
```
No need to clone in this case is there?

runfalk replied:
The object returns an `OpBuffer`, not a `&OpBuffer` like `buf` is. I am not sure we can take ownership within the function without changing the signature?

pedrocr replied:
Yeah, makes sense

pedrocr commented on src/decoders/iiq.rs:
```diff
@@ -54,7 +54,6 @@ impl<'a> Decoder for IiqDecoder<'a> {
     }
 
     let image = self.decode_compressed(data_offset, strip_offset, width, height);
-
```
another :)

pedrocr commented on src/decoders/dcr.rs:
```diff
@@ -41,7 +41,6 @@ impl<'a> Decoder for DcrDecoder<'a> {
     };
 
     let image = DcrDecoder::decode_kodak65000(src, &curve, width, height);
-
```
And one more :)

pedrocr commented on src/decoders/arw.rs:
```diff
@@ -73,7 +73,7 @@ impl<'a> Decoder for ArwDecoder<'a> {
 }
 
 impl<'a> ArwDecoder<'a> {
-  fn image_a100(&self, camera: &Camera) -> Result<RawImage,String> {
```
I wouldn't change the signature to these functions or to ok_camera. Passing the reference allows you to use it multiple times and sometimes that may be needed (e.g., first to decode, then to apply metadata).

runfalk replied:
Do you mean `ok_image`, not `ok_camera`? Previously the hash map in `RawLoader` ensured the instance was kept alive. I'm a bit inexperienced with lifetimes, but it seems to me to be problematic handle it, and making `ok_image` conuming it when it extracts its values seemed reasonable.

pedrocr replied:
ok_image right. All the values ok_image uses should be copyable or at least clonable so there should be no problem with just taking a reference and copying what you need and then discarding.

runfalk replied:
I can see one potential problem with this (maybe). The lifetime of these objects are tied to the lifetime of `RawLoader`, which makes sense when they were only tied to the camera toml file. Now they are based on orientation as well, which is image specific and still would have the life time of the `RawLoader` instance that created it. Wouldn't this mean a `RawLoader` would leak memory?

pedrocr replied:
It's very hard to make rust leak memory. In this case since orientation is Copy it won't even reference the original one. Even if it wasn't it would just mean ok_image would need to clone. Just leave all those signatures unchanged and the compiler will tell you if you need to do anything else.

pedrocr commented on src/imageops/transform.rs:
```diff
@@ -0,0 +1,64 @@
+    OrientationFlips { transpose: true, flip_x: true, flip_y: false } => flip_horizontal(&transpose(buf)),
+    OrientationFlips { transpose: true, flip_x: true, flip_y: true } => flip_vertical(&transpose(&flip_horizontal(buf))),
+  }
+}
```
Don't forget we also need some simple unit tests for each of these cases. This kind of code is too finicky to leave untested and it's easy enough to test. Do you have experience with doing those? I've done it once before in rust so if you're not comfortable I'll do them after merging.

runfalk replied:
I have done them once for my first unfinished pet project (https://github.com/runfalk/rustyspans/tree/master/src), but I'm not really sold on the currently recommended method of having them in the same file they are testing. I'm more for stuffing them in a separate file and/or folder to avoid unnecessary clutter. If you think you have a good idea of how to structure the tests, feel free to do it post merge.

I made a rough draft idea for how to test `OpBuffer` in a human readable way: https://gist.github.com/runfalk/7dda18b548d6b58fc59892e89648aa3b

pedrocr replied:
This seems neat. We can probably implement an actual equals operator so that normal assertions just work. I'll have a look at this after this merge.

runfalk replied:
I thought about implementing actual equals, `PartialEq`, but opted against since it will most likely only work in unit test scenarios. Implementing equality might suggest it is usable outside testing.

pedrocr commented on src/decoders/cr2.rs:
```diff
@@ -142,7 +142,8 @@ impl<'a> Decoder for Cr2Decoder<'a> {
 
-    let mut img = RawImage::new(camera, width, height, try!(self.get_wb(camera)), image);
+    let wb = self.get_wb(&camera)?;
+    let mut img = RawImage::new(camera, width, height, wb, image);
```
This seems like an unrelated code style change, no? Or do you need wb later for some reason?

runfalk replied:
I think it is because each argument is evaluated in order. The first one takes ownership of `camera` and then the white balance argument tried to borrow a reference, but it was already owned elsewhere. When reverting the change to make the first argument take a reference I can revert this line too.

pedrocr commented:
From the conversion we just had on IRC this is probably ready to merge once the whitespace changes are fixed. Do you have anything else you want to fix before I merge?

runfalk commented:
I've squashed all changes to a single commit that doesn't modify whitespace. I'm all for merging this and make other improvements in other pull requests.
